Library tests fail for projects which don't include @angular/animations module

We have a [library](https://angular.io/guide/creating-libraries#refactoring-parts-of-an-application-into-a-library) where in tests we use `ng-mocks`.  Our project doesn't include `@angular/animations` module as we don't use animations in our app. Our Angular is `12.2.13`

After updating `ng-mocks` from 11.x to latest 12.x (`12.5.0` at the moment of writing), following error started to appear when running library tests (`ng test <library-name>`), preventing tests from being run:

```
./node_modules/@angular/platform-browser/fesm2015/animations.js:9:0-83 - Error: Module not found: Error: Can't resolve '@angular/animations' in '/Users/vzozuliak/projects/dxng/node_modules/@angular/platform-browser/fesm2015'

./node_modules/@angular/platform-browser/fesm2015/animations.js:10:0-227 - Error: Module not found: Error: Can't resolve '@angular/animations/browser' in '/Users/vzozuliak/projects/dxng/node_modules/@angular/platform-browser/fesm2015'
```

With further investigation traced the issue to https://github.com/ike18t/ng-mocks/commit/ca65de3a2747fd1ac69c2dcef47a4896c7aa13a2 which was introduced in `12.1.0`. Double checked and re-assured that version `12.0.2` works well. 

Also, went to dist code of `12.5.0` and commented out this code in `ng-mocks/cjs/lib/mock-module/create-resolvers.js`, and again it proved to work just fine:

https://github.com/ike18t/ng-mocks/blob/17d798e74730fedff3b6af69f5fda70eecb89cc7/libs/ng-mocks/src/lib/mock-module/create-resolvers.ts#L19-L31

My assumption - Angular library builder is slightly different from app builder. I assume it tries to analyze all files for `require`s and to include them in it's own way, and so it tries to include  `@angular/platform-browser/animations` which then tries to include `@angular/animations` and so that fails, but since it all happens out of context of executing the actual file, it's not get caught with that `try/catch`

Workarounds:
- downgrade to `ng-mocks` 12.0.2 (not cool as we stuck with old version)
- write a script that patches `ng-mocks` commenting out the code above (not coll at all - monkey patching should be the last possible resort)

--- a/docs/articles/troubleshooting/browser-animations-module.md
+++ b/docs/articles/troubleshooting/browser-animations-module.md
@@ -4,38 +4,54 @@ description: Information and solutions how to test modules with BrowserAnimation
 sidebar_label: BrowserAnimationsModule
 ---
 
-By default, `ng-mocks` replaces `BrowserAnimationsModule` with `NoopAnimationsModule`.
+By default, Angular recommends replacing `BrowserAnimationsModule` with `NoopAnimationsModule` in unit tests.
 
-However, it can be changed via [`MockBuilder`](../api/MockBuilder.md) or [`ngMocks.guts`](../api/ngMocks/guts.md)
-when `NoopAnimationsModule` isn't a solution.
+In order to do so globally, you can use [`ngMocks.globalReplace`](../api/ngMocks/globalReplace.md):
+
+```ts title="src/test.ts"
+ngMocks.globalReplace(BrowserAnimationsModule, NoopAnimationsModule);
+```
+
+Now, every time `ng-mocks` sees `BrowserAnimationsModule`, it will substitute it with `NoopAnimationsModule`.
 
 ## MockBuilder
 
-```ts
-// No animations at all
-MockBuilder(MyComponent, MyModule).exclude(BrowserAnimationsModule);
+Please check how [`MockBuilder`](../api/MockBuilder.md) behaves in this case: 
 
-// Mock BrowserAnimationsModule
+```ts
+// BrowserAnimationsModule is replaced by NoopAnimationsModule.
+MockBuilder(MyComponent, MyModule);
+
+// BrowserAnimationsModule will be kept as it is.
+MockBuilder(MyComponent, MyModule).keep(BrowserAnimationsModule);
+
+// BrowserAnimationsModule will be mocked, not replaced.
 MockBuilder(MyComponent, MyModule).mock(BrowserAnimationsModule);
 
-// Keep BrowserAnimationsModule to test animations.
-MockBuilder(MyComponent, MyModule).keep(BrowserAnimationsModule);
+// BrowserAnimationsModule will be excluded from declarations.
+MockBuilder(MyComponent, MyModule).exclude(BrowserAnimationsModule);
 ```
 
 ## ngMocks.guts
 
+Please check how [`ngMocks.guts`](../api/ngMocks/guts.md) behaves in this case:
+
 ```ts
-// No animations at all
-ngMocks.guts(MyComponent, MyModule, BrowserAnimationsModule);
+// BrowserAnimationsModule is replaced by NoopAnimationsModule.
+ngMocks.guts(MyComponent, MyModule);
 
-// Mock BrowserAnimationsModule
-ngMocks.guts(MyComponent, [MyModule, BrowserAnimationsModule]);
-
-// Keep BrowserAnimationsModule to test animations.
+// BrowserAnimationsModule will be kept as it is.
 ngMocks.guts([MyComponent, BrowserAnimationsModule], MyModule);
+
+// BrowserAnimationsModule will be mocked, not replaced.
+ngMocks.guts([MyComponent, MyModule], BrowserAnimationsModule);
+
+// BrowserAnimationsModule will be excluded from declarations.
+ngMocks.guts(MyComponent, MyModule, BrowserAnimationsModule);
 ```
 
 ## fakeAsync
 
 A kept / mock `BrowserAnimationsModule` causes issues with `fakeAsync`.
-Please open an issue on github, if you have a case where `NoopAnimationsModule` isn't a solution.
+Please open an issue on [github](https://github.com/ike18t/ng-mocks/issues/new),
+if you have a case where `NoopAnimationsModule` isn't a solution.

--- a/e2e/a-min/docker-compose.yml
+++ b/e2e/a-min/docker-compose.yml
@@ -1,0 +1,22 @@
+version: '3'
+
+services:
+  install:
+    image: satantime/puppeteer-node:16.14.2
+    working_dir: /app
+    volumes:
+      - .:/app
+    command:
+      - npm
+      - install
+      - --force
+
+  test:
+    image: satantime/puppeteer-node:16.14.2
+    working_dir: /app
+    volumes:
+      - .:/app
+    command:
+      - npm
+      - run
+      - test:debug

--- a/e2e/a-min/package.json
+++ b/e2e/a-min/package.json
@@ -25,7 +25,6 @@
     "ng-mocks": "*"
   },
   "dependencies": {
-    "@angular/animations": "13.3.1",
     "@angular/common": "13.3.1",
     "@angular/compiler": "13.3.1",
     "@angular/core": "13.3.1",

--- a/libs/ng-mocks/src/lib/common/mock.ts
+++ b/libs/ng-mocks/src/lib/common/mock.ts
@@ -149,6 +149,7 @@ export class Mock {
     ngControl: any | null = null, // NgControl
   ) {
     const mockOf = (this.constructor as any).mockOf;
+    coreDefineProperty(this, '__ngMocks', true);
     coreDefineProperty(this, '__ngMocksInjector', injector);
     coreDefineProperty(this, '__ngMocksCtor', this.constructor);
     for (const key of this.__ngMocksConfig.queryScanKeys || /* istanbul ignore next */ []) {

--- a/libs/ng-mocks/src/lib/common/ng-mocks-global-overrides.ts
+++ b/libs/ng-mocks/src/lib/common/ng-mocks-global-overrides.ts
@@ -238,7 +238,7 @@ const install = () => {
     if (hooks.after.indexOf(resetTestingModule) === -1) {
       hooks.after.push(resetTestingModule);
     }
-    coreDefineProperty(TestBed, 'ngMocksOverridesInstalled', true, false);
+    coreDefineProperty(TestBed, 'ngMocksOverridesInstalled', true);
   }
 };
 

--- a/libs/ng-mocks/src/lib/mock-builder/mock-builder.promise.ts
+++ b/libs/ng-mocks/src/lib/mock-builder/mock-builder.promise.ts
@@ -73,6 +73,7 @@ export class MockBuilderPromise implements IMockBuilder {
   public build(): TestModuleMetadata {
     this.stash.backup();
     ngMocksUniverse.config.set('mockNgDefResolver', new Map());
+    ngMocksUniverse.flags.add('hasRootModule');
 
     try {
       const params = this.combineParams();
@@ -92,6 +93,7 @@ export class MockBuilderPromise implements IMockBuilder {
 
       return ngModule;
     } finally {
+      ngMocksUniverse.flags.delete('hasRootModule');
       ngMocksUniverse.config.delete('mockNgDefResolver');
       this.stash.restore();
     }

--- a/libs/ng-mocks/src/lib/mock-helper/mock-helper.console-throw.ts
+++ b/libs/ng-mocks/src/lib/mock-helper/mock-helper.console-throw.ts
@@ -8,7 +8,7 @@ const factory =
   (propName: string) =>
   (...args: any[]) => {
     const error = new Error(args.join(' '));
-    coreDefineProperty(error, 'ngMocksConsoleCatch', propName, false);
+    coreDefineProperty(error, 'ngMocksConsoleCatch', propName);
     throw error;
   };
 

--- a/libs/ng-mocks/src/lib/mock-helper/mock-helper.faster-install.ts
+++ b/libs/ng-mocks/src/lib/mock-helper/mock-helper.faster-install.ts
@@ -49,7 +49,7 @@ export default () => {
   if (!(TestBed as any).ngMocksFasterInstalled) {
     TestBed.configureTestingModule = configureTestingModule(TestBed.configureTestingModule);
     TestBed.resetTestingModule = resetTestingModule(TestBed.resetTestingModule);
-    coreDefineProperty(TestBed, 'ngMocksFasterInstalled', true, false);
+    coreDefineProperty(TestBed, 'ngMocksFasterInstalled', true);
   }
 
   return hooks;

--- a/libs/ng-mocks/src/lib/mock-module/create-resolvers.ts
+++ b/libs/ng-mocks/src/lib/mock-module/create-resolvers.ts
@@ -1,5 +1,4 @@
 import { Provider } from '@angular/core';
-import * as imports from '@angular/platform-browser/animations';
 
 import { isNgDef } from '../common/func.is-ng-def';
 import { isNgModuleDefWithProviders } from '../common/func.is-ng-module-def-with-providers';
@@ -11,21 +10,6 @@ import helperMockService from '../mock-service/helper.mock-service';
 
 import { MockModule } from './mock-module';
 
-// tslint:disable-next-line variable-name
-let BrowserAnimationsModule: any;
-// tslint:disable-next-line variable-name
-let NoopAnimationsModule: any;
-// istanbul ignore next
-let replaceWithNoop: (def: any) => boolean = () => false;
-
-BrowserAnimationsModule = imports.BrowserAnimationsModule;
-NoopAnimationsModule = imports.NoopAnimationsModule;
-replaceWithNoop = (def: any) =>
-  def === BrowserAnimationsModule &&
-  !!BrowserAnimationsModule &&
-  !!NoopAnimationsModule &&
-  !ngMocksUniverse.getResolution(def);
-
 const processDefMap: Array<[any, any]> = [
   ['c', MockComponent],
   ['d', MockDirective],
@@ -33,12 +17,6 @@ const processDefMap: Array<[any, any]> = [
 ];
 
 const processDef = (def: any) => {
-  // BrowserAnimationsModule is a very special case.
-  // If it is not resolved manually, we simply replace it with NoopAnimationsModule.
-  if (replaceWithNoop(def)) {
-    return NoopAnimationsModule;
-  }
-
   if (isNgDef(def, 'm') || isNgModuleDefWithProviders(def)) {
     return MockModule(def as any);
   }

--- a/libs/ng-mocks/src/lib/mock-module/mock-module.ts
+++ b/libs/ng-mocks/src/lib/mock-module/mock-module.ts
@@ -25,27 +25,50 @@ const flagReplace = (resolution?: string): boolean =>
 const flagNever = (ngModule?: any): boolean =>
   coreConfig.neverMockModule.indexOf(funcGetName(ngModule)) !== -1 && !ngMocksUniverse.flags.has('skipMock');
 
-const preprocessToggleFlag = (ngModule: Type<any>): boolean => {
+const preProcessFlags = (ngModule: Type<any>): { isRootModule: boolean; toggleSkipMockFlag: boolean } => {
   let toggleSkipMockFlag = false;
+  let isRootModule = true;
+
+  if (ngMocksUniverse.flags.has('hasRootModule')) {
+    isRootModule = false;
+  } else {
+    ngMocksUniverse.flags.add('hasRootModule');
+  }
 
   const resolution = ngMocksUniverse.getResolution(ngModule);
   if (flagMock(resolution)) {
     toggleSkipMockFlag = true;
     ngMocksUniverse.flags.delete('skipMock');
   }
-  if (flagKeep(resolution) || flagReplace(resolution) || flagNever(ngModule)) {
+  if (flagNever(ngModule)) {
+    toggleSkipMockFlag = true;
+    ngMocksUniverse.flags.add('skipMock');
+  }
+  if (!isRootModule && (flagKeep(resolution) || flagReplace(resolution))) {
     toggleSkipMockFlag = true;
     ngMocksUniverse.flags.add('skipMock');
   }
 
-  return toggleSkipMockFlag;
+  return {
+    isRootModule,
+    toggleSkipMockFlag,
+  };
 };
 
-const postprocessToggleFlag = (toggleSkipMockFlag: boolean): void => {
+const postProcessFlags = ({
+  isRootModule,
+  toggleSkipMockFlag,
+}: {
+  isRootModule: boolean;
+  toggleSkipMockFlag: boolean;
+}): void => {
   if (toggleSkipMockFlag && ngMocksUniverse.flags.has('skipMock')) {
     ngMocksUniverse.flags.delete('skipMock');
   } else if (toggleSkipMockFlag && !ngMocksUniverse.flags.has('skipMock')) {
     ngMocksUniverse.flags.add('skipMock');
+  }
+  if (isRootModule) {
+    ngMocksUniverse.flags.delete('hasRootModule');
   }
 };
 
@@ -73,7 +96,7 @@ const extractModuleAndProviders = (
   };
 };
 
-const getExistingMockModule = (ngModule: Type<any>): Type<any> | undefined => {
+const getExistingMockModule = (ngModule: Type<any>, isRootModule: boolean): Type<any> | undefined => {
   if (isMockNgDef(ngModule, 'm')) {
     return ngModule;
   }
@@ -85,6 +108,10 @@ const getExistingMockModule = (ngModule: Type<any>): Type<any> | undefined => {
   }
 
   // Now we check if we need to keep the original module or to replace it with some other.
+  // and there is no override in its resolution.
+  if (isRootModule || ngMocksUniverse.config.get('ngMocksDepsResolution')?.get(ngModule) === 'mock') {
+    return undefined;
+  }
   if (ngMocksUniverse.hasBuildDeclaration(ngModule)) {
     const instance = ngMocksUniverse.getBuildDeclaration(ngModule);
     if (isNgDef(instance, 'm') && instance !== ngModule) {
@@ -161,17 +188,20 @@ export function MockModule(module: any): any {
   funcImportExists(module, 'MockModule');
 
   const { ngModule, ngModuleProviders } = extractModuleAndProviders(module);
-  const toggleSkipMockFlag = preprocessToggleFlag(ngModule);
-  const mockModule = detectMockModule(ngModule, getExistingMockModule(ngModule));
-  // istanbul ignore else
-  if (ngMocksUniverse.flags.has('cacheModule')) {
-    ngMocksUniverse.cacheDeclarations.set(ngModule, mockModule);
-  }
-  if (ngMocksUniverse.flags.has('skipMock')) {
-    ngMocksUniverse.config.get('ngMocksDepsSkip')?.add(mockModule);
-  }
-  const mockModuleProviders = getMockProviders(ngModuleProviders);
-  postprocessToggleFlag(toggleSkipMockFlag);
+  const flags = preProcessFlags(ngModule);
+  try {
+    const mockModule = detectMockModule(ngModule, getExistingMockModule(ngModule, flags.isRootModule));
+    // istanbul ignore else
+    if (ngMocksUniverse.flags.has('cacheModule')) {
+      ngMocksUniverse.cacheDeclarations.set(ngModule, mockModule);
+    }
+    if (ngMocksUniverse.flags.has('skipMock')) {
+      ngMocksUniverse.config.get('ngMocksDepsSkip')?.add(mockModule);
+    }
+    const mockModuleProviders = getMockProviders(ngModuleProviders);
 
-  return generateReturn(module, ngModule, ngModuleProviders, mockModule, mockModuleProviders);
+    return generateReturn(module, ngModule, ngModuleProviders, mockModule, mockModuleProviders);
+  } finally {
+    postProcessFlags(flags);
+  }
 }

--- a/libs/ng-mocks/src/lib/mock-render/func.create-wrapper.ts
+++ b/libs/ng-mocks/src/lib/mock-render/func.create-wrapper.ts
@@ -103,7 +103,7 @@ export default (
   };
 
   ctor = generateWrapper({ ...meta, bindings, options });
-  coreDefineProperty(ctor, 'cacheKey', cacheKey, false);
+  coreDefineProperty(ctor, 'cacheKey', cacheKey);
   caches.unshift(ctor as any);
   caches.splice(ngMocksUniverse.global.get('mockRenderCacheSize') ?? coreConfig.mockRenderCacheSize);
 

--- a/libs/ng-mocks/src/lib/mock-render/func.install-prop-reader.ts
+++ b/libs/ng-mocks/src/lib/mock-render/func.install-prop-reader.ts
@@ -7,8 +7,8 @@ const createPropertyGet = (key: keyof any & string, reader: Record<keyof any, an
     if (typeof source[key] === 'function') {
       if (reader[`__ngMocks_${key}__origin`] !== source[key]) {
         const clone = helperMockService.createClone(source[key], reader, source);
-        coreDefineProperty(reader, `__ngMocks_${key}`, clone, false);
-        coreDefineProperty(reader, `__ngMocks_${key}__origin`, source[key], false);
+        coreDefineProperty(reader, `__ngMocks_${key}`, clone);
+        coreDefineProperty(reader, `__ngMocks_${key}__origin`, source[key]);
       }
 
       return reader[`__ngMocks_${key}`];
@@ -16,7 +16,7 @@ const createPropertyGet = (key: keyof any & string, reader: Record<keyof any, an
 
     return source[key];
   };
-  coreDefineProperty(handler, '__ngMocksProxy', true, false);
+  coreDefineProperty(handler, '__ngMocksProxy', true);
 
   return handler;
 };
@@ -31,7 +31,7 @@ const createPropertySet = (key: keyof any & string, reader: Record<keyof any, an
     }
     source[key] = newValue;
   };
-  coreDefineProperty(handler, '__ngMocksProxy', true, false);
+  coreDefineProperty(handler, '__ngMocksProxy', true);
 
   return handler;
 };
@@ -53,7 +53,7 @@ export default (
   if (!source) {
     return;
   }
-  coreDefineProperty(reader, '__ngMocks__source', source, false);
+  coreDefineProperty(reader, '__ngMocks__source', source);
   const exists = extractOwnKeys(reader);
   const fields = [...extractAllKeys(source), ...extra];
   for (const key of fields) {

--- a/libs/ng-mocks/src/lib/mock-render/mock-render-factory.ts
+++ b/libs/ng-mocks/src/lib/mock-render/mock-render-factory.ts
@@ -71,7 +71,7 @@ const fixtureMessage = [
 
 const handleFixtureError = (e: any) => {
   const error = new Error(fixtureMessage);
-  coreDefineProperty(error, 'parent', e, false);
+  coreDefineProperty(error, 'parent', e);
   throw error;
 };
 

--- a/libs/ng-mocks/src/lib/mock-service/helper.create-mock-from-prototype.ts
+++ b/libs/ng-mocks/src/lib/mock-service/helper.create-mock-from-prototype.ts
@@ -1,3 +1,4 @@
+import coreDefineProperty from '../common/core.define-property';
 import funcGetName from '../common/func.get-name';
 
 import helperMockService from './helper.mock-service';
@@ -6,6 +7,7 @@ import { MockedFunction } from './types';
 export default (service: any): { [key in keyof any]: MockedFunction } => {
   const mockName = funcGetName(service);
   const value: any = {};
+  coreDefineProperty(value, '__ngMocks', true);
 
   const methods = helperMockService.extractMethodsFromPrototype(service);
   for (const method of methods) {

--- a/tests/issue-1377/test.spec.ts
+++ b/tests/issue-1377/test.spec.ts
@@ -1,0 +1,425 @@
+// tslint:disable no-string-literal
+
+import {
+  state,
+  style,
+  transition,
+  trigger,
+} from '@angular/animations';
+import { AnimationDriver } from '@angular/animations/browser';
+import { Component, DebugElement, NgModule } from '@angular/core';
+import { TestBed } from '@angular/core/testing';
+import { BrowserModule } from '@angular/platform-browser';
+import {
+  BrowserAnimationsModule,
+  NoopAnimationsModule,
+} from '@angular/platform-browser/animations';
+import {
+  MockBuilder,
+  MockModule,
+  MockRender,
+  ngMocks,
+} from 'ng-mocks';
+
+@Component({
+  animations: [
+    trigger('openClose', [
+      state(
+        'open',
+        style({
+          opacity: 0,
+        }),
+      ),
+      state(
+        'closed',
+        style({
+          opacity: 1,
+        }),
+      ),
+      transition('open => closed', []),
+      transition('closed => open', []),
+    ]),
+  ],
+  selector: 'target',
+  template: `
+    <div [@openClose]="isOpen ? 'open' : 'closed'">
+      The box is now {{ isOpen ? 'Open' : 'Closed' }}!
+    </div>
+  `,
+})
+class TargetComponent {
+  public isOpen = true;
+
+  public toggle() {
+    this.isOpen = !this.isOpen;
+  }
+}
+
+@NgModule({
+  declarations: [TargetComponent],
+  exports: [TargetComponent],
+  imports: [BrowserModule, BrowserAnimationsModule],
+})
+class TargetModule {}
+
+const expectBrowserAnimations = async () => {
+  const fixture = MockRender(TargetComponent);
+  const driver = fixture.point.injector.get(AnimationDriver);
+  // the browser provider
+  // Skipping the check for IE and react-jsdom
+  if (
+    typeof (Element as any) !== 'undefined' &&
+    typeof (Element as any).prototype.animate === 'function'
+  ) {
+    expect(driver.constructor.name).not.toEqual(
+      'NoopAnimationDriver',
+    );
+  }
+  expect((driver as any).__ngMocks).toBeUndefined();
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Open!');
+    expect(div.nativeElement.style['opacity']).toEqual('0');
+  }
+
+  fixture.point.componentInstance.toggle();
+  fixture.detectChanges();
+  await fixture.whenStable();
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Closed!');
+    expect(div.nativeElement.style['opacity']).toEqual('1');
+  }
+};
+
+const expectNoopAnimations = async () => {
+  const fixture = MockRender(TargetComponent);
+  const driver = fixture.point.injector.get(AnimationDriver);
+  // the noop provider
+  expect(driver.constructor.name).toEqual('NoopAnimationDriver');
+  expect((driver as any).__ngMocks).toBeUndefined();
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Open!');
+    expect(div.nativeElement.style['opacity']).toEqual('0');
+  }
+
+  fixture.point.componentInstance.toggle();
+  fixture.detectChanges();
+  await fixture.whenStable();
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Closed!');
+    expect(div.nativeElement.style['opacity']).toEqual('1');
+  }
+};
+
+const expectNoAnimations = async () => {
+  const fixture = MockRender(TargetComponent);
+  const driver = fixture.point.injector.get(AnimationDriver);
+  // the mock provider
+  expect((driver as any).__ngMocks).toEqual(true);
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Open!');
+    expect(div.nativeElement.style['opacity']).toEqual('');
+  }
+
+  fixture.point.componentInstance.toggle();
+  fixture.detectChanges();
+  await fixture.whenStable();
+
+  {
+    const div = ngMocks.find<DebugElement>('div');
+    expect(ngMocks.formatText(div)).toContain('Closed!');
+    expect(div.nativeElement.style['opacity']).toEqual('');
+  }
+};
+
+const expectThrow = () => {
+  expect(() => MockRender(TargetComponent)).toThrowError(
+    /BrowserAnimationsModule/,
+  );
+};
+
+// @see https://github.com/ike18t/ng-mocks/issues/1377
+describe('issue-1377', () => {
+  describe('BrowserAnimationsModule', () => {
+    describe('classic', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule({
+          declarations: [TargetComponent],
+          imports: [BrowserModule, BrowserAnimationsModule],
+        }).compileComponents(),
+      );
+
+      it('renders animations', expectBrowserAnimations);
+    });
+
+    describe('MockModule', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule({
+          declarations: [TargetComponent],
+          imports: [
+            BrowserModule,
+            MockModule(BrowserAnimationsModule),
+          ],
+        }).compileComponents(),
+      );
+
+      it(
+        'does not render animations due to mocks',
+        expectNoAnimations,
+      );
+    });
+
+    describe('MockBuilder:default', () => {
+      beforeEach(() => MockBuilder(TargetComponent, TargetModule));
+
+      it(
+        'does not render animations due to mocks',
+        expectNoAnimations,
+      );
+    });
+
+    describe('MockBuilder:keep', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).keep(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it('renders animations due to .keep', expectBrowserAnimations);
+    });
+
+    describe('MockBuilder:mock', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).mock(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it(
+        'does not render animations due to mocks',
+        expectNoAnimations,
+      );
+    });
+
+    describe('MockBuilder:replace', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).replace(
+          BrowserAnimationsModule,
+          NoopAnimationsModule,
+        ),
+      );
+
+      it('uses noop animations', expectNoopAnimations);
+    });
+
+    describe('MockBuilder:exclude', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).exclude(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it('throws an error', expectThrow);
+    });
+
+    describe('ngMocks.guts:default', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(TargetComponent, TargetModule),
+        ).compileComponents(),
+      );
+
+      it(
+        'does not render animations due to mock',
+        expectNoAnimations,
+      );
+    });
+
+    describe('ngMocks.guts:mock', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts([TargetModule], BrowserAnimationsModule),
+        ).compileComponents(),
+      );
+
+      it(
+        'does not render animations due to mock',
+        expectNoAnimations,
+      );
+    });
+
+    describe('ngMocks.guts:keep', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(
+            [TargetComponent, BrowserAnimationsModule],
+            TargetModule,
+          ),
+        ).compileComponents(),
+      );
+
+      it('renders animations due to keep', expectBrowserAnimations);
+    });
+
+    describe('ngMocks.guts:exclude', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(
+            TargetComponent,
+            TargetModule,
+            BrowserAnimationsModule,
+          ),
+        ).compileComponents(),
+      );
+
+      it('throws an error', expectThrow);
+    });
+  });
+
+  describe('NoopAnimationsModule', () => {
+    beforeAll(() =>
+      ngMocks.globalReplace(
+        BrowserAnimationsModule,
+        NoopAnimationsModule,
+      ),
+    );
+    afterAll(() => ngMocks.globalWipe(BrowserAnimationsModule));
+
+    describe('classic', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule({
+          declarations: [TargetComponent],
+          imports: [BrowserModule, NoopAnimationsModule],
+        }).compileComponents(),
+      );
+
+      it('uses noop animations', expectNoopAnimations);
+    });
+
+    describe('MockModule', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule({
+          declarations: [TargetComponent],
+          imports: [
+            BrowserModule,
+            MockModule(BrowserAnimationsModule),
+          ],
+        }).compileComponents(),
+      );
+
+      it(
+        'does not render animations due to mocks',
+        expectNoAnimations,
+      );
+    });
+
+    describe('MockBuilder:default', () => {
+      beforeEach(() => MockBuilder(TargetComponent, TargetModule));
+
+      it('uses noop animations', expectNoopAnimations);
+    });
+
+    describe('MockBuilder:keep', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).keep(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it('renders animations due to .keep', expectBrowserAnimations);
+    });
+
+    describe('MockBuilder:mock', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).mock(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it(
+        'does not render animations due to .mock',
+        expectNoAnimations,
+      );
+    });
+
+    describe('MockBuilder:replace', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).replace(
+          BrowserAnimationsModule,
+          NoopAnimationsModule,
+        ),
+      );
+
+      it('uses noop animations', expectNoopAnimations);
+    });
+
+    describe('MockBuilder:exclude', () => {
+      beforeEach(() =>
+        MockBuilder(TargetComponent, TargetModule).exclude(
+          BrowserAnimationsModule,
+        ),
+      );
+
+      it('throws an error', expectThrow);
+    });
+
+    describe('ngMocks.guts:default', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(TargetComponent, TargetModule),
+        ).compileComponents(),
+      );
+
+      it('uses noop animations', expectNoopAnimations);
+    });
+
+    describe('ngMocks.guts:mock', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts([TargetModule], BrowserAnimationsModule),
+        ).compileComponents(),
+      );
+
+      it(
+        'does not render animations due to mock',
+        expectNoAnimations,
+      );
+    });
+
+    describe('ngMocks.guts:keep', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(
+            [TargetComponent, BrowserAnimationsModule],
+            TargetModule,
+          ),
+        ).compileComponents(),
+      );
+
+      it('renders animations due to keep', expectBrowserAnimations);
+    });
+
+    describe('ngMocks.guts:exclude', () => {
+      beforeEach(() =>
+        TestBed.configureTestingModule(
+          ngMocks.guts(
+            TargetComponent,
+            TargetModule,
+            BrowserAnimationsModule,
+          ),
+        ).compileComponents(),
+      );
+
+      it('throws an error', expectThrow);
+    });
+  });
+});

--- a/tests/issue-222/dom-shared-styles-host.spec.ts
+++ b/tests/issue-222/dom-shared-styles-host.spec.ts
@@ -8,7 +8,10 @@ import {
 import { Component, NgModule } from '@angular/core';
 import { TestBed } from '@angular/core/testing';
 import { BrowserModule } from '@angular/platform-browser';
-import { BrowserAnimationsModule } from '@angular/platform-browser/animations';
+import {
+  BrowserAnimationsModule,
+  NoopAnimationsModule,
+} from '@angular/platform-browser/animations';
 import { MockBuilder, MockRender, ngMocks } from 'ng-mocks';
 
 @Component({
@@ -55,7 +58,34 @@ class TargetComponent {
 })
 class TargetModule {}
 
+// Mocks BrowserAnimationsModule
+describe('issue-222:DomSharedStylesHost:classic', () => {
+  beforeEach(() =>
+    MockBuilder(TargetComponent, TargetModule).keep(
+      BrowserAnimationsModule,
+    ),
+  );
+
+  it('correctly handles DomSharedStylesHost in a mock module', () => {
+    const fixture = MockRender(TargetComponent);
+    expect(fixture.nativeElement.innerHTML).toContain(
+      'The box is now Open!',
+    );
+    // Animations are replaced with a mock copy, therefore no styles.
+    expect(fixture.nativeElement.innerHTML).toContain('yellow');
+  });
+});
+
+// Mocks BrowserAnimationsModule
 describe('issue-222:DomSharedStylesHost:mock', () => {
+  beforeAll(() =>
+    ngMocks.globalReplace(
+      BrowserAnimationsModule,
+      NoopAnimationsModule,
+    ),
+  );
+  afterAll(() => ngMocks.globalWipe(BrowserAnimationsModule));
+
   beforeEach(() =>
     MockBuilder(TargetComponent, TargetModule).mock(
       BrowserAnimationsModule,
@@ -72,7 +102,16 @@ describe('issue-222:DomSharedStylesHost:mock', () => {
   });
 });
 
+// Replaces BrowserAnimationsModule with Noop.
 describe('issue-222:DomSharedStylesHost:keep', () => {
+  beforeAll(() =>
+    ngMocks.globalReplace(
+      BrowserAnimationsModule,
+      NoopAnimationsModule,
+    ),
+  );
+  afterAll(() => ngMocks.globalWipe(BrowserAnimationsModule));
+
   beforeEach(() => MockBuilder(TargetComponent).keep(TargetModule));
 
   it('correctly handles DomSharedStylesHost in a kept module', () => {
@@ -85,6 +124,7 @@ describe('issue-222:DomSharedStylesHost:keep', () => {
   });
 });
 
+// Mocks BrowserAnimationsModule as any other module
 describe('issue-222:DomSharedStylesHost:guts', () => {
   beforeEach(() =>
     TestBed.configureTestingModule(

--- a/tests/issue-641/test.spec.ts
+++ b/tests/issue-641/test.spec.ts
@@ -48,6 +48,14 @@ class TargetComponent {
 class TargetModule {}
 
 describe('issue-641', () => {
+  beforeAll(() =>
+    ngMocks.globalReplace(
+      BrowserAnimationsModule,
+      NoopAnimationsModule,
+    ),
+  );
+  afterAll(() => ngMocks.globalWipe(BrowserAnimationsModule));
+
   describe('BrowserAnimationsModule:default', () => {
     beforeEach(() => MockBuilder(TargetComponent, TargetModule));
 

